src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp

```diff
@@ -189,44 +192,108 @@ namespace OpenLoco::World::MapGenerator
         return ((randVal & 0xFF) * landObj->numVariations) >> 8;
     }
 
+    static void applySurfaceStyleToMarkedTiles(HeightMap& heightMap, uint8_t surfaceStyle, bool requireMark)
+    {
+        for (auto& pos : World::getDrawableTileRange())
+        {
+            const bool tileIsMarked = heightMap.isMarkerSet({ pos.x, pos.y });
+            if (requireMark != tileIsMarked)
+                continue;
+
+            auto tile = TileManager::get(pos);
+            auto* surface = tile.surface();
+            if (surface == nullptr)
+                continue;
+
+            surface->setTerrain(surfaceStyle);
+            auto res = getRandomTerrainVariation(*surface);
+            if (res)
+                surface->setVariation(*res);
+        }
+    }
+
     // 0x0046A379
     static void generateTerrainFarFromWater(HeightMap& heightMap, uint8_t surfaceStyle)
     {
-        _heightMap = heightMap.data();
-        registers regs;
-        regs.ebx = surfaceStyle;
-        call(0x0046A379, regs);
-        _heightMap = nullptr;
+        heightMap.resetMarkerFlags();
+
+        // Mark tiles near water
+        auto seaLevel = getGameState().seaLevel;
+        for (auto pos : getWorldRange())
+        {
+            auto height = heightMap.getHeight({ pos.x, pos.y });
+            if (height > seaLevel)
+                continue;
+
+            for (auto lookaheadPos : getClampedRange(pos, pos + TilePos2(50, 50)))
```

Comment:
i don't think this is correct. I think it is meant to be pos - (25, 25) to pos + (25,25)